runtime/runtime/src/lib.rs

```
@@ -1499,55 +1667,116 @@ impl Runtime {
+
+        all_processed.sort_by_key(|pt| pt.index);
+
+        for processed in all_processed {
+            metrics::TRANSACTION_PROCESSED_TOTAL.inc();
+            match processed.result {
+                PerTransactionResult::Success {
+                    validated_tx,
+                    verification_result,
+                    outcome,
+                    receipt,
+                } => {
+                    match safe_add_balance(
+                        processing_state.stats.balance.tx_burnt_amount,
+                        verification_result.burnt_amount,
                     ) {
-                        Ok(outcome) => outcome,
+                        Ok(new_balance) => {
+                            processing_state.stats.balance.tx_burnt_amount = new_balance;
+                        }
                         Err(err) => {
+                            metrics::TRANSACTION_PROCESSED_FAILED_TOTAL.inc();
                             tracing::debug!(
                                 target: "runtime",
-                                ?tx_hash,
+                                tx_hash=?validated_tx.get_hash(),
                                 ?err,
-                                "invalid transaction ignored (process_transaction error)",
+                                "invalid transaction ignored (burnt gas overflow)",
                             );
                             continue;
                         }
-                    };
-                    if receipt.receiver_id() == validated_tx.signer_id() {
-                        processing_state.local_receipts.push_back(receipt);
+                    }
+
+                    metrics::TRANSACTION_PROCESSED_SUCCESSFULLY_TOTAL.inc();
+                    if receipt.receiver_id() == validated_tx.to_tx().signer_id() {
+                        processing_state.local_receipts.push_back(*receipt);
                     } else {
                         receipt_sink.forward_or_buffer_receipt(
```

Original comment:
> Why not ?-propagate it as it was in the previous code? In general if apply returns Err, the chunk is basically considered invalid.

Done. It looks like we only return errors on various over/under-flows here, probably meaning some logic error elsewhere.

---

So the distinction here is that if we return an error, then the whole chunk becomes invalid. Alternatively we drop some transactions that are generally invalid (e.g. expired) thus keeping the chunk valid still.

So it is important that the behaviour between the old code and the code after these changes keeps returning an error where it used to return one and keeps ignoring transactions for errors for which the error was ignored before.

Does that make sense?

---

Hm, so two angles here, still. Some of the overflow errors here aren't necessarily a defect in *this* code, but rather an issue with how chunks got produced, making them invalid. That holds for things like accumulating total gas/compute cost in the chunk.

On the other hand here in `forward_or_buffer_receipt` I agree that we might be very well looking at a local code defect (i.e. an issue in implementation of *this* code.) In which case some sort of a more prominent failure mode would be more appropriate, I guess. But this PR is not the right place to hash it out anyhow. Consider filing an issue with your discovery so we can look at it later (or just eventually close the issue anyway, I guess…)

---

Yeah, no point creating the issue to only close it a year from now. Resolving.